### PR TITLE
mcs: add option to control redirect http request

### LINKED ISSUE #7690: Support using an option to decide if the request should be redirected to scheduling service
## Enhancement Task
It's hard to know if the rule is right since we can only get the rules from the scheduling service when it is enabled. We also need the real rules on the PD side so that we can know if there is a problem in synchronization through comparison.

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #7690

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: tools/pd-ctl/pdctl/command/config_command.go [hunk 1]
```diff
@@ -437,6 +438,7 @@ func NewPlacementRulesCommand() *cobra.Command {
 	show.Flags().String("id", "", "rule id")
 	show.Flags().String("region", "", "region id")
 	show.Flags().Bool("detail", false, "detailed match info for region")
+	show.Flags().Bool(flagRedirectMicroService, true, "allow to redirect to micro service")
```
Comment (rleungx) [inline tools/pd-ctl/pdctl/command/config_command.go:441]: What if the flag is misused in non-micro service env?

File: pkg/utils/apiutil/serverapi/middleware.go [hunk 2]
```diff
@@ -119,6 +119,9 @@ func (h *redirector) matchMicroServiceRedirectRules(r *http.Request) (bool, stri
 	if len(h.microserviceRedirectRules) == 0 {
 		return false, ""
 	}
+	if r.URL.Query().Get("redirect_mcs") == "false" {
```
Comment (rleungx) [inline pkg/utils/apiutil/serverapi/middleware.go:122]: redirect_to_ms?

File: tools/pd-ctl/pdctl/command/config_command.go [hunk 3]
```diff
@@ -27,24 +27,26 @@ import (
...
+	clusterVersionPrefix     = "pd/api/v1/config/cluster-version"
+	rulesPrefix              = "pd/api/v1/config/rules"
+	rulesBatchPrefix         = "pd/api/v1/config/rules/batch"
+	rulePrefix               = "pd/api/v1/config/rule"
+	ruleGroupPrefix          = "pd/api/v1/config/rule_group"
+	ruleGroupsPrefix         = "pd/api/v1/config/rule_groups"
+	replicationModePrefix    = "pd/api/v1/config/replication-mode"
+	ruleBundlePrefix         = "pd/api/v1/config/placement-rule"
+	pdServerPrefix           = "pd/api/v1/config/pd-server"
+	flagRedirectMicroService = "redirect_to_mcs"
```
Comment (rleungx) [inline tools/pd-ctl/pdctl/command/config_command.go:49]: Shall we make it easier to use?

File: tools/pd-ctl/tests/config/config_test.go [hunk 4]
```diff
@@ -343,6 +343,48 @@ func (suite *configTestSuite) checkConfig(cluster *pdTests.TestCluster) {
...
+	re.NoError(failpoint.Disable("github.com/tikv/pd/pkg/dashboard/adapter/skipDashboardLoop"))
+}
+
+func (suite *configTestSuite) checkConfigForwardControl(cluster *pdTests.TestCluster) {
+	re := suite.Require()
+	leaderServer := cluster.GetLeaderServer()
+	pdAddr := leaderServer.GetAddr()
+	cmd := ctl.GetRootCmd()
+
+	args := []string{"-u", pdAddr, "config", "show", "--from_api_server"}
```
Comment (rleungx) [inline tools/pd-ctl/tests/config/config_test.go:359]: How do we know if this flag is effective?

File: pkg/utils/apiutil/serverapi/middleware.go [hunk 5]
```diff
@@ -200,11 +202,8 @@ func (h *redirector) ServeHTTP(w http.ResponseWriter, r *http.Request, next http
 			return
 		}
 		clientUrls = append(clientUrls, targetAddr)
-		// Add a header to the response, this is not a failure injection
-		// it is used for testing, to check whether the request is forwarded to the micro service
-		failpoint.Inject("checkHeader", func() {
-			w.Header().Set(apiutil.ForwardToMicroServiceHeader, "true")
-		})
+		// Add a header to the response, it is used to mark whether the request has been forwarded to the micro service.
+		w.Header().Add(apiutil.XForwardedToMicroServiceHeader, "true")
```
Comment (CabinfeverB) [inline pkg/utils/apiutil/serverapi/middleware.go:206]: Can we not put this header on every forward request? This reduces network traffic. For example, when the `len(r.Header.Get(apiutil.XForbiddenForwardToMicroServiceHeader)) > 0`, that's the same as making an extra mark, similar to failpoint.

Comment (CabinfeverB) [inline pkg/utils/apiutil/serverapi/middleware.go:206]: Or can we can keep failpoint?  If `len(r.Header.Get(apiutil.XForbiddenForwardToMicroServiceHeader)) > 0` or failpoint enabled, we add the header into resp.

Comment (lhy1024) [reply pkg/utils/apiutil/serverapi/middleware.go:206]: If `len(r.Header.Get(apiutil.XForbiddenForwardToMicroServiceHeader)) > 0` , code will not reach here

Comment (CabinfeverB) [reply pkg/utils/apiutil/serverapi/middleware.go:206]: If it is as you say, then the new check in `pdctl/global.go` is also unreachable

File: tools/pd-ctl/tests/config/config_test.go (hunk 4 above)
Comment (lhy1024) [reply tools/pd-ctl/tests/config/config_test.go]: I have added more tests to check

File: pkg/utils/apiutil/serverapi/middleware.go (hunk 5 above)
Comment (lhy1024) [reply pkg/utils/apiutil/serverapi/middleware.go:206]: But if we use curl or pdctl to call api, how do we tell if it's coming from the scheduling server or the api server?

**Review by rleungx — APPROVED** [review]

File: tools/pd-ctl/tests/config/config_test.go [hunk 6]
```diff
@@ -343,6 +343,220 @@ func (suite *configTestSuite) checkConfig(cluster *pdTests.TestCluster) {
...
+		re.Len(ruleManager.GetAllRules(), 2)
+	}
+
+	// show placement rules
+	testRules(true /*isFromAPIServer*/, "show")
+	testRules(false /*isFromAPIServer*/, "show")
+
+	// load placement rules
+	testRules(false /*isFromAPIServer*/, "load", "--out="+fname)
+	testRules(false /*isFromAPIServer*/, "load", "--out="+fname)
```
Comment (rleungx) [inline tools/pd-ctl/tests/config/config_test.go:545]: What is the difference?

File: tools/pd-ctl/tests/config/config_test.go [hunk 7]
```diff
@@ -343,6 +343,220 @@ func (suite *configTestSuite) checkConfig(cluster *pdTests.TestCluster) {
...
+
+	// show placement rules group
+	testRules(true /*isFromAPIServer*/, "rule-group", "show", placement.DefaultGroupID)
+	testRules(false /*isFromAPIServer*/, "rule-group", "show", placement.DefaultGroupID)
```
Comment (rleungx) [inline tools/pd-ctl/tests/config/config_test.go:549]: If we need to test both true and false, why do we need this flag?

**Review by CabinfeverB — APPROVED** [review]

**Pull request merged**